File: src/lib/store.ts
import { create } from "zustand";
import { persist, createJSONStorage } from "zustand/middleware";
import { useEffect } from "react";
import { deviceLang, resolveLang, type Lang, type LangPref } from "./i18n";
import { applyTheme, type ColorSchemeId } from "./theme";
import { isoOf, todayIso } from "./dates";

export type CalendarEvent = {
  id: string;
  title: string;
  notes?: string;
  date: string;
  endDate?: string;
  startTime?: string;
  endTime?: string;
  allDay?: boolean;
  reminderDate?: string;
  reminderTime?: string;
  done?: boolean;
};

export type TabId = "today" | "months" | "events" | "weather" | "more";

const DEFAULT_CITIES = ["phnom-penh", "banteay-meanchey", "kampong-cham", "kratie"] as const;
const SEED_TITLES = new Set(["ប្រជុំក្រុមការងារ", "ធ្វើបុណ្យទាន", "ពិធីគ្រួសារ", "ផ្សារទំនើប"]);
const SEED_IDS = new Set(["seed-meeting", "seed-offering", "seed-family"]);

export function isSeed(e: CalendarEvent) {
  return SEED_IDS.has(e.id) || SEED_TITLES.has(e.title);
}

type ThemeMode = "light" | "dark" | "system";

type State = {
  events: CalendarEvent[];
  cursor: string;
  selected: string;
  theme: ThemeMode;
  colorScheme: ColorSchemeId;
  materialYou: boolean;
  extraDark: boolean;
  accentColor: string;
  highlightColor: string;
  highlightAlpha: number;
  lang: Lang;
  langPref: LangPref;
  setupDone: boolean;
  lastTab: TabId;
  lastEventsPane: "holidays" | "tasks";
  weatherCities: string[];
  installed: boolean;
  notifyOn: boolean;
  backgroundOn: boolean;
  locationOn: boolean;
  autoLaunchOn: boolean;
  weekStartsOn: 0 | 1 | 6;
  hydrated: boolean;
  addEvent: (e: CalendarEvent) => void;
  updateEvent: (e: CalendarEvent) => void;
  deleteEvent: (id: string) => void;
  toggleEventDone: (id: string) => void;
  setCursor: (iso: string) => void;
  setSelected: (iso: string) => void;
  goToDate: (iso: string) => void;
  setTheme: (t: ThemeMode) => void;
  setColorScheme: (id: ColorSchemeId) => void;
  setMaterialYou: (v: boolean) => void;
  setExtraDark: (v: boolean) => void;
  setAccentColor: (v: string) => void;
  setHighlightColor: (v: string) => void;
  setHighlightAlpha: (v: number) => void;
  setLang: (v: LangPref) => void;
  setSetupDone: (v: boolean) => void;
  setLastTab: (v: TabId) => void;
  setLastEventsPane: (v: "holidays" | "tasks") => void;
  addWeatherCity: (id: string) => void;
  removeWeatherCity: (id: string) => void;
  setInstalled: (v: boolean) => void;
  setNotifyOn: (v: boolean) => void;
  setBackgroundOn: (v: boolean) => void;
  setLocationOn: (v: boolean) => void;
  setAutoLaunchOn: (v: boolean) => void;
  setWeekStartsOn: (v: 0 | 1 | 6) => void;
  clearEventReminders: () => void;
  resetWeatherCities: () => void;
  resetAppData: () => void;
  paintTheme: () => void;
};

function paint(s: State) {
  applyTheme({
    theme: s.theme,
    colorScheme: s.colorScheme,
    materialYou: s.materialYou,
    extraDark: s.extraDark,
    accentColor: s.accentColor,
    highlightColor: s.highlightColor,
    highlightAlpha: s.highlightAlpha,
  });
  if (typeof document !== "undefined") {
    document.documentElement.lang = s.lang;
    document.title = s.lang === "en" ? "Khmer Calendar" : "ប្រតិទិនខ្មែរ";
  }
}

const memory = () => {
  const m = new Map<string, string>();
  return {
    getItem: (k: string) => m.get(k) ?? null,
    setItem: (k: string, v: string) => {
      m.set(k, v);
    },
    removeItem: (k: string) => {
      m.delete(k);
    },
  };
};

function storage() {
  if (typeof window === "undefined") return memory();
  try {
    const t = "__khmer_cal_probe";
    localStorage.setItem(t, "1");
    localStorage.removeItem(t);
    return localStorage;
  } catch {
    return memory();
  }
}

export const useStore = create<State>()(
  persist(
    (set, get) => ({
      events: [],
      cursor: todayIso(),
      selected: todayIso(),
      theme: "system",
      colorScheme: "rose",
      materialYou: true,
      extraDark: false,
      accentColor: "#F5C400",
      highlightColor: "#FF3B30",
      highlightAlpha: 0.22,
      lang: deviceLang(),
      langPref: "auto",
      setupDone: false,
      lastTab: "months",
      lastEventsPane: "holidays",
      weatherCities: [...DEFAULT_CITIES],
      installed: false,
      notifyOn: false,
      backgroundOn: false,
      locationOn: false,
      autoLaunchOn: false,
      weekStartsOn: 1,
      hydrated: false,
      addEvent: (e) => set((s) => ({ events: [...s.events, e] })),
      updateEvent: (e) => set((s) => ({ events: s.events.map((x) => (x.id === e.id ? e : x)) })),
      deleteEvent: (id) => set((s) => ({ events: s.events.filter((x) => x.id !== id) })),
      toggleEventDone: (id) =>
        set((s) => ({ events: s.events.map((x) => (x.id === id ? { ...x, done: !x.done } : x)) })),
      setCursor: (iso) => set({ cursor: iso }),
      setSelected: (iso) => set({ selected: iso }),
      goToDate: (iso) => set({ cursor: iso, selected: iso }),
      setTheme: (theme) => {
        set({ theme });
        paint(get());
      },
      setColorScheme: (colorScheme) => {
        set({ colorScheme, materialYou: true });
        paint(get());
      },
      setMaterialYou: (materialYou) => {
        set({ materialYou });
        paint(get());
      },
      setExtraDark: (extraDark) => {
        set({ extraDark });
        paint(get());
      },
      setAccentColor: (accentColor) => {
        set({ accentColor });
        paint(get());
      },
      setHighlightColor: (highlightColor) => {
        set({ highlightColor });
        paint(get());
      },
      setHighlightAlpha: (highlightAlpha) => {
        set({ highlightAlpha: Math.min(1, Math.max(0, highlightAlpha)) });
        paint(get());
      },
      setLang: (pref) => {
        const lang = resolveLang(pref);
        set({ langPref: pref, lang });
        paint(get());
      },
      setSetupDone: (setupDone) => set({ setupDone }),
      setLastTab: (lastTab) => set({ lastTab }),
      setLastEventsPane: (lastEventsPane) => set({ lastEventsPane }),
      addWeatherCity: (id) =>
        set((s) => (s.weatherCities.includes(id) ? s : { weatherCities: [...s.weatherCities, id] })),
      removeWeatherCity: (id) =>
        set((s) => ({ weatherCities: s.weatherCities.filter((c) => c !== id) })),
      setInstalled: (installed) => set({ installed }),
      setNotifyOn: (notifyOn) => set({ notifyOn }),
      setBackgroundOn: (backgroundOn) => set({ backgroundOn }),
      setLocationOn: (locationOn) => set({ locationOn }),
      setAutoLaunchOn: (autoLaunchOn) => set({ autoLaunchOn }),
      setWeekStartsOn: (weekStartsOn) => set({ weekStartsOn }),
      clearEventReminders: () =>
        set((s) => ({
          events: s.events.map((e) => ({ ...e, reminderDate: "", reminderTime: "" })),
        })),
      resetWeatherCities: () => set({ weatherCities: [...DEFAULT_CITIES] }),
      resetAppData: () => {
        set({
          events: [],
          theme: "system",
          colorScheme: "rose",
          materialYou: true,
          extraDark: false,
          accentColor: "#F5C400",
          highlightColor: "#FF3B30",
          highlightAlpha: 0.22,
          lastEventsPane: "holidays",
          weatherCities: [...DEFAULT_CITIES],
          installed: false,
          notifyOn: false,
          backgroundOn: false,
          locationOn: false,
          autoLaunchOn: false,
          weekStartsOn: 1,
        });
        paint(get());
      },
      paintTheme: () => paint(get()),
    }),
    {
      name: "khmer-calendar-v3",
      version: 18,
      storage: createJSONStorage(() => storage()),
      skipHydration: true,
      partialize: (s) => ({
        events: s.events.filter((e) => !isSeed(e)),
        theme: s.theme,
        colorScheme: s.colorScheme,
        materialYou: s.materialYou,
        extraDark: s.extraDark,
        accentColor: s.accentColor,
        highlightColor: s.highlightColor,
        highlightAlpha: s.highlightAlpha,
        lang: s.lang,
        langPref: s.langPref,
        setupDone: s.setupDone,
        lastTab: s.lastTab,
        lastEventsPane: s.lastEventsPane,
        weatherCities: s.weatherCities,
        installed: s.installed,
        notifyOn: s.notifyOn,
        backgroundOn: s.backgroundOn,
        locationOn: s.locationOn,
        autoLaunchOn: s.autoLaunchOn,
        weekStartsOn: s.weekStartsOn,
      }),
      merge: (persisted, current) => {
        const p = (persisted ?? {}) as Partial<State>;
        const events = (p.events ?? []).filter((e) => !isSeed(e));
        const langPref = p.langPref ?? current.langPref;
        return {
          ...current,
          ...p,
          events,
          langPref,
          lang: resolveLang(langPref),
        };
      },
      onRehydrateStorage: () => (state) => {
        if (!state) {
          useStore.setState({ hydrated: true });
          return;
        }
        state.events = (state.events ?? []).filter((e) => !isSeed(e));
        state.lang = resolveLang(state.langPref ?? "auto");
        state.hydrated = true;
        state.paintTheme();
      },
    },
  ),
);

export function useHydrateStore() {
  useEffect(() => {
    const api = useStore.persist;
    const finish = () => {
      if (!useStore.getState().hydrated) useStore.setState({ hydrated: true });
      useStore.getState().paintTheme();
    };
    if (api.hasHydrated()) {
      finish();
      return;
    }
    const unsub = api.onFinishHydration(finish);
    void api.rehydrate();
    const t = window.setTimeout(finish, 400);
    return () => {
      unsub();
      window.clearTimeout(t);
    };
  }, []);
}

export function newId() {
  return `evt-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 7)}`;
}

export function eventsOn(events: CalendarEvent[], iso: string) {
  return events.filter((e) => e.date === iso);
}

export { todayIso, isoOf };

File: khmer_calendar/lib/store.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'dates.dart';
import 'i18n.dart';
import 'theme.dart';

class CalendarEvent {
  CalendarEvent({
    required this.id,
    required this.title,
    this.notes,
    required this.date,
    this.endDate,
    this.startTime,
    this.endTime,
    this.allDay,
    this.reminderDate,
    this.reminderTime,
    this.done,
  });

  String id;
  String title;
  String? notes;
  String date;
  String? endDate;
  String? startTime;
  String? endTime;
  bool? allDay;
  String? reminderDate;
  String? reminderTime;
  bool? done;

  CalendarEvent copy() => CalendarEvent(
        id: id,
        title: title,
        notes: notes,
        date: date,
        endDate: endDate,
        startTime: startTime,
        endTime: endTime,
        allDay: allDay,
        reminderDate: reminderDate,
        reminderTime: reminderTime,
        done: done,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'notes': notes,
        'date': date,
        'endDate': endDate,
        'startTime': startTime,
        'endTime': endTime,
        'allDay': allDay,
        'reminderDate': reminderDate,
        'reminderTime': reminderTime,
        'done': done,
      };

  factory CalendarEvent.fromJson(Map<String, dynamic> j) => CalendarEvent(
        id: j['id'] as String,
        title: j['title'] as String? ?? '',
        notes: j['notes'] as String?,
        date: j['date'] as String? ?? '',
        endDate: j['endDate'] as String?,
        startTime: j['startTime'] as String?,
        endTime: j['endTime'] as String?,
        allDay: j['allDay'] as bool?,
        reminderDate: j['reminderDate'] as String?,
        reminderTime: j['reminderTime'] as String?,
        done: j['done'] as bool?,
      );
}

enum TabId { today, months, events, weather, more }

class RouterTick extends ChangeNotifier {
  void bump() => notifyListeners();
}

const _legacyDefaultCities = ['phnom-penh', 'banteay-meanchey', 'kampong-cham', 'kratie'];

class AppStore extends ChangeNotifier {
  List<CalendarEvent> events = [];
  String cursor = todayIso();
  String selected = todayIso();
  String theme = 'system';
  ColorSchemeId colorScheme = ColorSchemeId.slate;
  bool materialYou = true;
  bool dynamicColor = false;
  bool extraDark = false;
  String accentColor = '#F5C400';
  String highlightColor = '#FF3B30';
  double highlightAlpha = 0.22;
  String langPref = 'auto';
  Lang lang = deviceLang();
  bool setupDone = false;
  TabId lastTab = TabId.months;
  String lastEventsPane = 'holidays';
  List<String> weatherCities = [];
  bool installed = false;
  bool notifyOn = false;
  bool backgroundOn = false;
  bool locationOn = false;
  bool autoLaunchOn = false;
  bool notifyTasks = true;
  bool notifyDaily = false;
  bool notifySil = true;
  bool notifyPublic = true;
  bool notifyOthers = true;
  int weekStartsOn = 1;
  bool hydrated = false;
  String? pendingRoute;
  Timer? _persistDebounce;
  final routerTick = RouterTick();

  Brightness get brightness {
    if (theme == 'light') return Brightness.light;
    if (theme == 'dark') return Brightness.dark;
    return WidgetsBinding.instance.platformDispatcher.platformBrightness;
  }

  Future<void> hydrate() async {
    final started = DateTime.now();
    try {
      final prefs = await SharedPreferences.getInstance();
      final raw = prefs.getString('khmer-calendar-v4') ?? prefs.getString('khmer-calendar-v3');
      if (raw != null) {
        final p = jsonDecode(raw) as Map<String, dynamic>;
        events = ((p['events'] as List?) ?? [])
            .whereType<Map>()
            .map((e) => CalendarEvent.fromJson(Map<String, dynamic>.from(e)))
            .toList();
        theme = p['theme'] as String? ?? theme;
        final scheme = p['colorScheme'] as String?;
        if (scheme != null) {
          colorScheme = ColorSchemeId.values.firstWhere(
            (s) => s.name == scheme,
            orElse: () => ColorSchemeId.slate,
          );
        }
        materialYou = p['materialYou'] as bool? ?? true;
        dynamicColor = p['dynamicColor'] as bool? ?? false;
        extraDark = p['extraDark'] as bool? ?? false;
        accentColor = p['accentColor'] as String? ?? accentColor;
        highlightColor = p['highlightColor'] as String? ?? highlightColor;
        highlightAlpha = (p['highlightAlpha'] as num?)?.toDouble() ?? highlightAlpha;
        langPref = p['langPref'] as String? ?? 'auto';
        setupDone = p['setupDone'] as bool? ?? false;
        final tab = p['lastTab'] as String?;
        if (tab != null) {
          lastTab = TabId.values.firstWhere((t) => t.name == tab, orElse: () => TabId.months);
        }
        lastEventsPane = p['lastEventsPane'] as String? ?? 'holidays';
        weatherCities = ((p['weatherCities'] as List?) ?? []).cast<String>();
        if (listEquals(weatherCities, _legacyDefaultCities)) weatherCities = [];
        installed = p['installed'] as bool? ?? false;
        notifyOn = p['notifyOn'] as bool? ?? false;
        backgroundOn = p['backgroundOn'] as bool? ?? false;
        locationOn = p['locationOn'] as bool? ?? false;
        autoLaunchOn = p['autoLaunchOn'] as bool? ?? false;
        notifyTasks = p['notifyTasks'] as bool? ?? true;
        notifyDaily = p['notifyDaily'] as bool? ?? false;
        notifySil = p['notifySil'] as bool? ?? true;
        final oldHolidays = p['notifyHolidays'] as bool? ?? true;
        notifyPublic = p['notifyPublic'] as bool? ?? oldHolidays;
        notifyOthers = p['notifyOthers'] as bool? ?? p['notifyReligious'] as bool? ?? oldHolidays;
        weekStartsOn = p['weekStartsOn'] as int? ?? 1;
      }
    } catch (_) {
      /* first run */
    }
    lang = resolveLang(langPref);
    IntlHelper.localeName = lang == Lang.km ? 'km' : 'en';
    final wait = 720 - DateTime.now().difference(started).inMilliseconds;
    if (wait > 0) await Future<void>.delayed(Duration(milliseconds: wait));
    hydrated = true;
    notifyListeners();
    routerTick.bump();
  }

  Future<void> persist() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      'khmer-calendar-v4',
      jsonEncode({
        'events': events.map((e) => e.toJson()).toList(),
        'theme': theme,
        'colorScheme': colorScheme.name,
        'materialYou': materialYou,
        'dynamicColor': dynamicColor,
        'extraDark': extraDark,
        'accentColor': accentColor,
        'highlightColor': highlightColor,
        'highlightAlpha': highlightAlpha,
        'langPref': langPref,
        'setupDone': setupDone,
        'lastTab': lastTab.name,
        'lastEventsPane': lastEventsPane,
        'weatherCities': weatherCities,
        'installed': installed,
        'notifyOn': notifyOn,
        'backgroundOn': backgroundOn,
        'locationOn': locationOn,
        'autoLaunchOn': autoLaunchOn,
        'notifyTasks': notifyTasks,
        'notifyDaily': notifyDaily,
        'notifySil': notifySil,
        'notifyPublic': notifyPublic,
        'notifyOthers': notifyOthers,
        'weekStartsOn': weekStartsOn,
      }),
    );
  }

  void _touch({bool save = true}) {
    notifyListeners();
    if (!save) return;
    _persistDebounce?.cancel();
    _persistDebounce = Timer(const Duration(milliseconds: 180), persist);
  }

  void addEvent(CalendarEvent e) {
    events = [...events, e];
    _touch();
  }

  void updateEvent(CalendarEvent e) {
    events = events.map((x) => x.id == e.id ? e : x).toList();
    _touch();
  }

  void deleteEvent(String id) {
    events = events.where((x) => x.id != id).toList();
    _touch();
  }

  void toggleEventDone(String id) {
    events = events.map((x) {
      if (x.id != id) return x;
      final c = x.copy();
      c.done = !(x.done ?? false);
      return c;
    }).toList();
    _touch();
  }

  void setCursor(String iso) {
    cursor = iso;
    _touch(save: false);
  }

  void setSelected(String iso) {
    selected = iso;
    _touch(save: false);
  }

  void goToDate(String iso) {
    cursor = iso;
    selected = iso;
    _touch(save: false);
  }

  String? takePendingRoute() {
    final route = pendingRoute;
    pendingRoute = null;
    return route;
  }

  void applyLaunch({String? tab, String? date}) {
    final iso = (date ?? '').trim();
    if (iso.isNotEmpty) goToDate(iso);
    final key = (tab ?? '').trim();
    String? route;
    switch (key) {
      case 'today':
      case 'day':
        lastTab = TabId.today;
        route = '/day';
      case 'months':
        lastTab = TabId.months;
        route = '/months';
      case 'events':
        lastTab = TabId.events;
        route = '/events';
      case 'weather':
        lastTab = TabId.weather;
        route = '/weather';
      case 'more':
        lastTab = TabId.more;
        route = '/more';
      default:
        if (iso.isNotEmpty) {
          lastTab = TabId.today;
          route = '/day';
        }
    }
    if (route != null) pendingRoute = route;
    _touch();
    routerTick.bump();
  }

  void setTheme(String v) {
    theme = v;
    _touch();
  }

  void setColorScheme(ColorSchemeId id) {
    colorScheme = id;
    materialYou = true;
    _touch();
  }

  void setMaterialYou(bool v) {
    materialYou = v;
    _touch();
  }

  void setDynamicColor(bool v) {
    dynamicColor = v;
    _touch();
  }

  void setExtraDark(bool v) {
    extraDark = v;
    _touch();
  }

  void setAccentColor(String v) {
    accentColor = v;
    _touch();
  }

  void setHighlightColor(String v) {
    highlightColor = v;
    _touch();
  }

  void setHighlightAlpha(double v) {
    highlightAlpha = v.clamp(0, 1);
    _touch();
  }

  void setLang(String pref) {
    langPref = pref;
    lang = resolveLang(pref);
    IntlHelper.localeName = lang == Lang.km ? 'km' : 'en';
    _touch();
  }

  void setSetupDone(bool v) {
    if (setupDone == v) return;
    setupDone = v;
    _touch();
    routerTick.bump();
  }

  void setLastTab(TabId v) {
    lastTab = v;
    _touch();
  }

  void setLastEventsPane(String v) {
    lastEventsPane = v;
    _touch();
  }

  bool addWeatherCity(String id) {
    if (weatherCities.contains(id)) return false;
    weatherCities = [...weatherCities, id];
    _touch();
    return true;
  }

  void removeWeatherCity(String id) {
    weatherCities = weatherCities.where((c) => c != id).toList();
    _touch();
  }

  void setNotifyOn(bool v) {
    if (notifyOn == v) return;
    notifyOn = v;
    _touch();
  }

  void setBackgroundOn(bool v) {
    if (backgroundOn == v) return;
    backgroundOn = v;
    _touch();
  }

  void setLocationOn(bool v) {
    if (locationOn == v) return;
    locationOn = v;
    _touch();
  }

  void setAutoLaunchOn(bool v) {
    if (autoLaunchOn == v) return;
    autoLaunchOn = v;
    _touch();
  }

  void setNotifyPublic(bool v) {
    if (notifyPublic == v) return;
    notifyPublic = v;
    _touch();
  }

  void setNotifyOthers(bool v) {
    if (notifyOthers == v) return;
    notifyOthers = v;
    _touch();
  }

  void setNotifyTasks(bool v) {
    if (notifyTasks == v) return;
    notifyTasks = v;
    _touch();
  }

  void setNotifyDaily(bool v) {
    if (notifyDaily == v) return;
    notifyDaily = v;
    _touch();
  }

  void setNotifySil(bool v) {
    if (notifySil == v) return;
    notifySil = v;
    _touch();
  }

  void setWeekStartsOn(int v) {
    weekStartsOn = v;
    _touch();
  }

  void setInstalled(bool v) {
    installed = v;
    _touch();
  }

  void clearEventReminders() {
    events = events.map((e) {
      final c = e.copy();
      c.reminderDate = '';
      c.reminderTime = '';
      return c;
    }).toList();
    _touch();
  }

  void resetWeatherCities() {
    weatherCities = [];
    _touch();
  }

  void resetAppData() {
    events = [];
    theme = 'system';
    colorScheme = ColorSchemeId.slate;
    materialYou = true;
    dynamicColor = false;
    extraDark = false;
    accentColor = '#F5C400';
    highlightColor = '#FF3B30';
    highlightAlpha = 0.22;
    lastEventsPane = 'holidays';
    weatherCities = [];
    installed = false;
    notifyOn = false;
    backgroundOn = false;
    locationOn = false;
    autoLaunchOn = false;
    notifyTasks = true;
    notifyDaily = false;
    notifySil = true;
    notifyPublic = true;
    notifyOthers = true;
    weekStartsOn = 1;
    _touch();
  }
}
